### PR TITLE
Cross-Chain Queries (ICS-031) refactoring

### PR DESCRIPTION
**Pull request by SangyunOck**
## Description

Complementary pr for #2723 

After I created the earlier one, there were pretty a lot of changes such as refactoring for crates such as cosmos/ibc-proto-rs and relayer_types. 

For better readability and easy code management, opened another one synced up to the latest master branch.

Merkle proof of query result via abci_query is included in this version which is the main difference between the earlier one but pre-requisite is needed because of the mechanism of stride binary like adding specific address to admin list, so e2e test using hermes' test framework is not included currently.

Instead, I uploaded some scripts for local test:
https://github.com/validance/stride-gaia-local

Following README.md might enough to understand how the kv query works!

Thank you
______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

**Review by ljoss17 — COMMENTED** [review]
Great job! I left a few initial comments, and I will continue to review this.

File: crates/relayer-types/src/applications/ics31_icq/events.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    block_events: &BTreeMap<String, Vec<String>>,
+    key: &str,
+) -> Result<String, Error> {
+    let res = block_events
+        .get(key)
+        .ok_or_else(Vec::new)
+        .map_err(|_: Vec<&String>| Error::parse())?
+        .get(0)
+        .ok_or(Err(()))
+        .map_err(|_: Result<&String, ()>| Error::parse())?;
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer-types/src/applications/ics31_icq/events.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/136201af7533b79623ef9c94fd3b8748e2afe517

File: crates/relayer-types/src/applications/ics31_icq/events.rs [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+                block_events,
+                &format!("{}.{}", EVENT_TYPE_PREFIX, "action"),
+            )?,
+            query_id: fetch_first_element_from_events(
+                block_events,
+                &format!("{}.{}", EVENT_TYPE_PREFIX, "query_id"),
+            )?,
+            chain_id: ChainId::from_string(&chain_id_str),
+            connection_id: ConnectionId::from_str(&connection_id_str)
+                .map_err(|_| Error::parse())?,
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer-types/src/applications/ics31_icq/events.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/845881790af92a3a774333d7ee72553e6a6371a4

File: crates/relayer/src/chain/handle/counting.rs [hunk 3]
```diff
@@ -483,4 +484,11 @@ impl<Handle: ChainHandle> ChainHandle for CountingChainHandle<Handle> {
         self.inner
             .maybe_register_counterparty_payee(channel_id, port_id, counterparty_payee)
     }
+
+    fn cross_chain_query(
+        &self,
+        request: Vec<CrossChainQueryRequest>,
+    ) -> Result<Vec<CrossChainQueryResponse>, Error> {
+        self.inner.cross_chain_query(request)
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer/src/chain/handle/counting.rs:493]: https://github.com/informalsystems/hermes/pull/2915/commits/23e9b1e0cf76a5a902a85425b3578a385ad5621a

File: crates/relayer/src/object.rs [hunk 4]
```diff
@@ -198,6 +214,7 @@ impl Object {
             Object::Channel(c) => &c.src_chain_id == src_chain_id,
             Object::Packet(p) => &p.src_chain_id == src_chain_id,
             Object::Wallet(_) => false,
+            Object::CrossChainQuery(_) => true,
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer/src/object.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/96591ecff50680fc945515dbb7af229e49214464

File: crates/relayer/src/worker/cross_chain_query.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+            let client_a = ForeignClient::find(
+                chain_b_handle.clone(),
+                chain_a_handle.clone(),
+                connection_end.client_id(),
+            )
+            .map_err(|_| TaskError::Fatal(RunError::query()))?;
+
+            let target_height = Height::new(
+                chain_b_handle.id().version(),
+                cross_chain_query_responses.get(0).unwrap().height as u64,
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer/src/worker/cross_chain_query.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/1e8e5dc63e5c848db2eee373eb279b983350ebd8

File: crates/relayer-types/src/applications/ics31_icq/events.rs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            query_type,
+            height: Height::from_str(&height_str).map_err(|_| Error::parse())?,
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer-types/src/applications/ics31_icq/events.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/845881790af92a3a774333d7ee72553e6a6371a4

File: crates/relayer/src/event/monitor.rs [hunk 7]
```diff
@@ -156,6 +156,10 @@ pub mod queries {
     pub fn ibc_channel() -> Query {
         Query::eq("message.module", "ibc_channel")
     }
+
+    pub fn ibc_query() -> Query {
```
Comment (ljoss17) [inline crates/relayer/src/event/monitor.rs:160]: This should be added to the list of events to query, https://github.com/informalsystems/hermes/blob/master/crates/relayer/src/event/monitor.rs#L134-L141.

File: crates/relayer-types/src/applications/ics31_icq/events.rs [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    }
+}
+
+fn fetch_first_element_from_events(
+    block_events: &BTreeMap<String, Vec<String>>,
+    key: &str,
+) -> Result<String, Error> {
+    let res = block_events
+        .get(key)
+        .ok_or_else(|| Error::event(format!("key not found: {}", key)))?
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer-types/src/applications/ics31_icq/events.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/5c3637d14dbf80f277784610f338d909562ee965

File: crates/relayer-types/src/applications/ics31_icq/events.rs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+        .get(0)
+        .ok_or_else(|| Error::event(format!("attribute not found for key: {}", key)))?;
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer-types/src/applications/ics31_icq/events.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/5c3637d14dbf80f277784610f338d909562ee965

File: crates/relayer/src/event/monitor.rs (hunk 7 above)
Comment (SangyunOck) [reply crates/relayer/src/event/monitor.rs:161]: https://github.com/informalsystems/hermes/pull/2915/commits/2be623da2cb287692330c51a5bb6562b4ef570d7

**Review by ljoss17 — COMMENTED** [review]
Some small nits, but great work!!

File: crates/relayer/src/worker/cross_chain_query.rs [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+                                    .get_signer()
+                                    .map_err(|_| TaskError::Fatal(RunError::query()))?,
+                            )
+                            .map_err(|_| TaskError::Fatal(RunError::query()))?,
+                    );
+                }
+
+                chain_a_handle
+                    .send_messages_and_wait_check_tx(TrackedMsgs::new_uuid(
+                        chain_a_msgs.clone(),
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer/src/worker/cross_chain_query.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/5344cf3b53e8dbaacf32959bee7d7a3982bebc20

File: crates/relayer-types/src/applications/ics31_icq/error.rs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+            | e | { format!("Tendermint error: {:?}", e.error) },
+
+        Query
+            | _ | { "Failed to query data" },
+
+        Proof
+            | _ | { "Proof not found" },
+
+        ProtoEncode
+            | _ | { "Failed to encode " },
```
[the remark this answers is not in the document]
Comment (SangyunOck) [reply crates/relayer-types/src/applications/ics31_icq/error.rs]: https://github.com/informalsystems/hermes/pull/2915/commits/11c57c4ad4fbdbb602ed7ed9235713399c8b8786

**Review by ljoss17 — APPROVED** [review]

**Pull request merged**